File: cubist/get-schemas.js
const fs = require("fs");
const api = require("./openapi.json");

const RELEVANT_PATHS = [
  ["/v0/org/{org_id}/keys/{key_id}", ["get"]],
  ["/v1/org/{org_id}/blob/sign/{key_id}"],
  ["/v1/org/{org_id}/token/refresh"],
];

const getComponentKey = (ref) => {
  const path = ref.split("/");

  const componentName = path.pop();
  const componentType = path.pop();

  return [componentType, componentName];
};

const getComponent = (components, [componentType, componentName]) => {
  return components[componentType][componentName];
};

const assignComponent = (components, newComponents, key) => {
  newComponents[key[0]][key[1]] = getComponent(components, key);
};

const getRefs = (obj, refs = []) => {
  if (typeof obj !== "object" || !obj) return;

  if (obj.$ref) {
    refs.push(obj.$ref);
  }

  Object.values(obj).forEach((maybeObj) => {
    getRefs(maybeObj, refs);
  });

  return;
};

const components = api.components;
const newComponents = { ...api.components, schemas: {}, responses: {} };

let refs = [];

// get all relevant refs from paths
RELEVANT_PATHS.forEach(([path, methods = []]) => {
  if (methods.length === 0) {
    getRefs(api.paths[path], refs);
  } else {
    methods.forEach((method) => {
      getRefs(api.paths[path][method], refs);
    });
  }
});

// repeat
while (refs.length !== 0) {
  // get all relevant components from refs
  // and assign them to newComponents
  refs.forEach((ref) => {
    assignComponent(components, newComponents, getComponentKey(ref));
  });

  // reset refs to []
  refs = [];

  // get all relevant refs from newComponents.(schemas|responses)
  getRefs(newComponents.schemas, refs);
  getRefs(newComponents.responses, refs);

  // filter refs such that all refs that are already in newComponents are removed
  refs = refs.filter((ref) => {
    const [componentType, componentName] = getComponentKey(ref);
    return !newComponents[componentType][componentName];
  });
}

api.components = newComponents;
api.paths = RELEVANT_PATHS.reduce((acc, [path, methods = []]) => {
  if (methods.length === 0) {
    acc[path] = api.paths[path];
  } else {
    acc[path] = methods.reduce((acc, method) => {
      acc[method] = api.paths[path][method];
      return acc;
    }, {});
  }

  return acc;
}, {});

fs.writeFileSync("./filtered-openapi.json", JSON.stringify(api, null, 2));

File: cubist/get-schemas.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

var relevantPaths = []string{
	"/v1/org/{org_id}/blob/sign/{key_id}",
	"/v1/org/{org_id}/token/refresh",
	"/v0/org/{org_id}/keys/{key_id}",
}

func getComponentKey(ref string) (string, string) {
	parts := strings.Split(ref, "/")
	componentName := parts[len(parts)-1]
	componentType := parts[len(parts)-2]
	return componentType, componentName
}

func getComponent(components map[string]map[string]interface{}, componentType, componentName string) interface{} {
	return components[componentType][componentName]
}

func getRefs(obj interface{}) []string {
	refs := []string{}

	switch v := obj.(type) {
	case map[string]interface{}:
		if ref, ok := v["$ref"].(string); ok {
			refs = append(refs, ref)
		}
		for _, val := range v {
			refs = append(refs, getRefs(val)...)
		}
	case []interface{}:
		for _, item := range v {
			refs = append(refs, getRefs(item)...)
		}
	}

	return refs
}

func main() {
	// Read OpenAPI JSON
	data, err := os.ReadFile("openapi.json")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	var api map[string]interface{}
	if err := json.Unmarshal(data, &api); err != nil {
		fmt.Println("Error parsing JSON:", err)
		return
	}

	components := api["components"].(map[string]interface{})
	typedComponents := make(map[string]map[string]interface{})
	for key, value := range components {
		typedComponents[key] = value.(map[string]interface{})
	}
	newComponents := map[string]interface{}{
		"schemas":   map[string]interface{}{},
		"responses": map[string]interface{}{},
	}

	// Copy existing components
	for k, v := range components {
		newComponents[k] = v
	}

	refs := []string{}
	paths := api["paths"].(map[string]interface{})

	// Get initial refs from relevant paths
	for _, path := range relevantPaths {
		refs = append(refs, getRefs(paths[path])...)
	}

	// Resolve and collect referenced components
	for len(refs) > 0 {
		currentRefs := refs
		refs = []string{}

		for _, ref := range currentRefs {
			componentType, componentName := getComponentKey(ref)

			// Ensure component type exists in newComponents
			if _, ok := newComponents[componentType]; !ok {
				newComponents[componentType] = map[string]interface{}{}
			}
			// Add component if not already present
			if _, ok := newComponents[componentType].(map[string]interface{})[componentName]; !ok {
				newComponents[componentType].(map[string]interface{})[componentName] =
					getComponent(typedComponents, componentType, componentName)
			}
		}

		// Get new refs from schemas and responses
		if schemas, ok := newComponents["schemas"].(map[string]interface{}); ok {
			refs = append(refs, getRefs(schemas)...)
		}
		if responses, ok := newComponents["responses"].(map[string]interface{}); ok {
			refs = append(refs, getRefs(responses)...)
		}

		// Filter out already processed refs
		filteredRefs := []string{}
		for _, ref := range refs {
			componentType, componentName := getComponentKey(ref)
			if _, ok := newComponents[componentType].(map[string]interface{})[componentName]; !ok {
				filteredRefs = append(filteredRefs, ref)
			}
		}
		refs = filteredRefs
	}

	// Update API spec
	api["components"] = newComponents

	// Filter paths
	filteredPaths := map[string]interface{}{}
	for _, path := range relevantPaths {
		filteredPaths[path] = paths[path]
	}
	api["paths"] = filteredPaths

	// Write filtered OpenAPI spec
	output, err := json.MarshalIndent(api, "", "  ")
	if err != nil {
		fmt.Println("Error marshaling JSON:", err)
		return
	}

	if err := os.WriteFile("filtered-openapi.json", output, 0644); err != nil {
		fmt.Println("Error writing file:", err)
	}
}
